vanilla proving is way too slow

### Description

*NOTE: this issue is marked **P0** because it blocks validation of large-sector sealing.*

Vanilla proving takes far longer than it should. Here is example output from benchmarks I ran last week:

```console
➜  rust-proofs git:(zigzag-example-taper) ✗ ./target/release/examples/zigzag --m 5 --expansion 8 --layers 10 --challenges 5 --size 262144 --groth
…
Feb 22 22:49:03.378 INFO vanilla_proving_time: 80.99321579 seconds, target: stats, place: filecoin-proofs/examples/zigzag.rs:208 zigzag, root: filecoin-proofs
```

Why does it take 80 seconds to generate 750 inclusion proofs?
  - 750 = 50 challenges * 15 inclusion proofs. (15 proofs = data node + replica node + 13 parent nodes)
  - This is (more or less) 5 x 15 = 75 proofs from each of 10 layers
  - In this example (256 MiB), proofs are 24 hashes in length.
  - Each hash is 32 bytes.
  - So total proof size is 750 x 24 x 32 = 600,000 bytes.
  - The trees are already generated, so no hashing should be happening.
  - The encodings are stored with the trees, so no replication should be happening.
  - Why is it taking 80 seconds to process 600,000 bytes?

It gets worse.

When I try to generate proofs after replicating 64 GiB, the vanilla proving never terminates. Using 8 partitions, I see 8 processes pinged 100% utilization.

In this case:
 - I was generating 8,848 x 15 = 132,720 proofs.
 -  For 64 GiB, the proofs are 32 hashes (32 bytes each).
 - So total data is 132,720 x 32 x 32 = 135,905,280
 - This is ~226 times larger than the previous (slow) example.
 - So at the previous rate, this should have taken about 5 hours.
 - It did not complete in 5 hours.

So… something is wrong. Let's find the problem and fix it.

### Acceptance criteria

Vanilla proofs for 1GiB take dramatically less than 80 seconds to generate. Maybe a couple seconds?

### Risks + pitfalls

This could be annoying to debug if it requires repeated performance of slow replication.
  - Use blake2s instead of pedersen. But first test both and see whether the choice affects proving time. If it does, that's unexpected and probably means hashing is happening (which it shouldn't).
  - Try to use small sectors. However, we do know the problem seems to get worse as sizes grow, so that may be difficult.
  - It would be easier if the products of replication could be saved to disk so we could restart at the proving phase.
    - This should be part of the disk-backed merkle-tree generation story (#481).
    -  If it's not, consider adding — both in support of this but *also especially* for ongoing benchmarks/experiments on large data. (cc: @schomatis)
    -  Having the ability to restart mid-replication, or to restart an interrupted replication would be extremely valuable.
    - This will also probably be a necessary eventual production feature.
    - Finally, it might be the case that making this possible fixes the problem. In case the problem is somehow related to memory bloat incurred during replication.

### Where to begin

- Run the command in the example above and record initial vanilla proving times.
- Read the code.
- Profile.
- Acquire but do not necessarily sacrifice a chicken.

I took the first baby steps toward removing the offending clones [here](https://github.com/filecoin-project/rust-fil-proofs/compare/bugfix/slow-vanilla-proving?expand=1): 

 However, we're immediately faced with this:

```rust
    Checking storage-proofs v0.1.0 (/Users/clwk/dev/go/src/github.com/filecoin-project/rust-proofs/storage-proofs)
error[E0508]: cannot move out of type `[merkle_light::merkle::MerkleTree<<<Self as layered_drgporep::Layers>::Hasher as hasher::types::Hasher>::Domain, <<Self as layered_drgporep::Layers>::Hasher as hasher::types::Hasher>::Function>]`, a non-copy slice
   --> storage-proofs/src/layered_drgporep.rs:268:29
    |
268 |                     tree_d: aux[layer],
    |                             ^^^^^^^^^^ cannot move out of here

error[E0508]: cannot move out of type `[merkle_light::merkle::MerkleTree<<<Self as layered_drgporep::Layers>::Hasher as hasher::types::Hasher>::Domain, <<Self as layered_drgporep::Layers>::Hasher as hasher::types::Hasher>::Function>]`, a non-copy slice
   --> storage-proofs/src/layered_drgporep.rs:269:29
    |
269 |                     tree_r: aux[layer + 1],
    |                             ^^^^^^^^^^^^^^ cannot move out of here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0508`.
error: Could not compile `storage-proofs`.

To learn more, run the command again with --verbose.
```

There are two obvious paths here:
  - Make the fields take references and borrow here.
  - Ensure ownership of `aux` so we can move.

I started down each path enough to know it's likely to get hairy and propagate through the code base.

This explains why the offending clones have survived this long, but the performance problem introduced by cloning these trees is a show-stopper. So the time has come, the clones must die.

Let the clone wars begin, and may they be as bloodless as possible.

![image](https://user-images.githubusercontent.com/1746729/54077689-f70b6400-4270-11e9-9669-4e39e1a4989e.png)

--- a/filecoin-proofs/examples/drgporep-vanilla-disk.rs
+++ b/filecoin-proofs/examples/drgporep-vanilla-disk.rs
@@ -89,7 +89,10 @@ fn do_the_work<H: Hasher>(data_size: usize, m: usize, sloth_iter: usize, challen
         tau: Some(tau),
     };
 
-    let priv_inputs = PrivateInputs::<H> { aux: &aux };
+    let priv_inputs = PrivateInputs::<H> {
+        tree_d: &aux.tree_d,
+        tree_r: &aux.tree_r,
+    };
 
     param_duration += start.elapsed();
     let samples: u32 = 30;

--- a/filecoin-proofs/examples/drgporep-vanilla.rs
+++ b/filecoin-proofs/examples/drgporep-vanilla.rs
@@ -102,7 +102,10 @@ fn do_the_work<H: Hasher>(data_size: usize, m: usize, sloth_iter: usize, challen
         tau: Some(tau),
     };
 
-    let priv_inputs = PrivateInputs::<H> { aux: &aux };
+    let priv_inputs = PrivateInputs::<H> {
+        tree_d: &aux.tree_d,
+        tree_r: &aux.tree_r,
+    };
 
     param_duration += start.elapsed();
     let samples: u32 = 30;

--- a/filecoin-proofs/examples/zigzag.rs
+++ b/filecoin-proofs/examples/zigzag.rs
@@ -278,6 +278,7 @@ fn do_the_work<H: 'static>(
         let vanilla_proving = start.elapsed();
         total_proving += vanilla_proving;
 
+        info!(FCP_LOG, "vanilla_proving_time: {:?}", vanilla_proving; "target" => "stats");
         if dump_proofs {
             dump_proof_bytes(&all_partition_proofs);
         }

--- a/storage-proofs/src/circuit/drgporep.rs
+++ b/storage-proofs/src/circuit/drgporep.rs
@@ -524,7 +524,10 @@ mod tests {
             challenges: vec![challenge],
             tau: Some(tau.into()),
         };
-        let priv_inputs = drgporep::PrivateInputs::<PedersenHasher> { aux: &aux };
+        let priv_inputs = drgporep::PrivateInputs::<PedersenHasher> {
+            tree_d: &aux.tree_d,
+            tree_r: &aux.tree_r,
+        };
 
         let proof_nc =
             drgporep::DrgPoRep::<PedersenHasher, _>::prove(&pp, &pub_inputs, &priv_inputs)
@@ -686,7 +689,10 @@ mod tests {
             challenges,
             tau: Some(tau),
         };
-        let private_inputs = drgporep::PrivateInputs { aux: &aux };
+        let private_inputs = drgporep::PrivateInputs {
+            tree_d: &aux.tree_d,
+            tree_r: &aux.tree_r,
+        };
 
         // This duplication is necessary so public_params don't outlive public_inputs and private_inputs.
         let setup_params = compound_proof::SetupParams {

--- a/storage-proofs/src/drgporep.rs
+++ b/storage-proofs/src/drgporep.rs
@@ -24,7 +24,8 @@ pub struct PublicInputs<T: Domain> {
 
 #[derive(Debug)]
 pub struct PrivateInputs<'a, H: 'a + Hasher> {
-    pub aux: &'a porep::ProverAux<H>,
+    pub tree_d: &'a MerkleTree<H::Domain, H::Function>,
+    pub tree_r: &'a MerkleTree<H::Domain, H::Function>,
 }
 
 #[derive(Debug)]
@@ -270,8 +271,8 @@ where
             let challenge = pub_inputs.challenges[i] % pub_params.graph.size();
             assert_ne!(challenge, 0, "cannot prove the first node");
 
-            let tree_d = &priv_inputs.aux.tree_d;
-            let tree_r = &priv_inputs.aux.tree_r;
+            let tree_d = &priv_inputs.tree_d;
+            let tree_r = &priv_inputs.tree_r;
             let domain_replica = tree_r.as_slice();
 
             let data = domain_replica[challenge];
@@ -641,7 +642,10 @@ mod tests {
                 tau: Some(tau.clone().into()),
             };
 
-            let priv_inputs = PrivateInputs::<H> { aux: &aux };
+            let priv_inputs = PrivateInputs::<H> {
+                tree_d: &aux.tree_d,
+                tree_r: &aux.tree_r,
+            };
 
             let real_proof = DrgPoRep::<H, _>::prove(&pp, &pub_inputs, &priv_inputs).unwrap();
 

--- a/storage-proofs/src/drgporep.rs
+++ b/storage-proofs/src/drgporep.rs
@@ -308,16 +308,14 @@ where
                 // )?;
 
                 let extracted = decode_domain_block::<H>(
-                    pub_params.graph.degree(),
                     pub_params.sloth_iter,
                     &pub_inputs.replica_id,
                     domain_replica,
                     challenge,
-                    parents,
-                )?
-                .into_bytes();
+                    &parents,
+                )?;
                 data_nodes.push(DataProof {
-                    data: H::Domain::try_from_bytes(&extracted)?,
+                    data: extracted,
                     proof: MerkleProof::new_from_proof(&node_proof),
                 });
             }

--- a/storage-proofs/src/layered_drgporep.rs
+++ b/storage-proofs/src/layered_drgporep.rs
@@ -270,10 +270,8 @@ pub trait Layers {
                 let inner_layers = layers - layer;
 
                 let new_priv_inputs = drgporep::PrivateInputs {
-                    aux: &porep::ProverAux {
-                        tree_d: aux[layer].clone(),
-                        tree_r: aux[layer + 1].clone(),
-                    },
+                    tree_d: &aux[layer],
+                    tree_r: &aux[layer + 1],
                 };
                 let layer_diff = total_layers - inner_layers;
 

--- a/storage-proofs/src/vde.rs
+++ b/storage-proofs/src/vde.rs
@@ -98,13 +98,13 @@ pub fn decode_domain_block<H>(
 where
     H: Hasher,
 {
-    let key = create_key_domain::<H>(replica_id, node, parents, data)?;
+    let key = create_domain_key::<H>(replica_id, node, parents, data)?;
 
     Ok(H::sloth_decode(&key, &data[node], sloth_iter))
 }
 
 /// Creates the encoding key, using domain encoded data.
-fn create_key_domain<H: Hasher>(
+fn create_domain_key<H: Hasher>(
     id: &H::Domain,
     node: usize,
     parents: &[usize],

--- a/storage-proofs/src/vde.rs
+++ b/storage-proofs/src/vde.rs
@@ -85,31 +85,46 @@ where
     let key = create_key::<H>(replica_id, v, &parents, &data, graph.degree())?;
     let node_data = H::Domain::try_from_bytes(&data_at_node(data, v)?)?;
 
-    // TODO: round constant
     Ok(H::sloth_decode(&key, &node_data, sloth_iter))
 }
 
-pub fn decode_domain_block<'a, H>(
-    degree: usize,
+pub fn decode_domain_block<H>(
     sloth_iter: usize,
-    replica_id: &'a H::Domain,
-    data: &'a [H::Domain],
-    v: usize,
-    parents: Vec<usize>,
+    replica_id: &H::Domain,
+    data: &[H::Domain],
+    node: usize,
+    parents: &[usize],
 ) -> Result<H::Domain>
 where
     H: Hasher,
 {
-    let byte_data = data
-        .iter()
-        .flat_map(H::Domain::into_bytes)
-        .collect::<Vec<u8>>();
+    let key = create_key_domain::<H>(replica_id, node, parents, data)?;
 
-    let key = create_key::<H>(replica_id, v, &parents, &byte_data, degree)?;
-    let node_data = data[v];
+    Ok(H::sloth_decode(&key, &data[node], sloth_iter))
+}
 
-    // TODO: round constant
-    Ok(H::sloth_decode(&key, &node_data, sloth_iter))
+/// Creates the encoding key, using domain encoded data.
+fn create_key_domain<H: Hasher>(
+    id: &H::Domain,
+    node: usize,
+    parents: &[usize],
+    data: &[H::Domain],
+) -> Result<H::Domain> {
+    let mut hasher = Blake2s::new().hash_length(32).to_state();
+    hasher.update(id.as_ref());
+
+    for parent in parents.iter() {
+        // special super shitty case
+        // TODO: unsuck
+        if node == parents[0] {
+            // skip, as we would only write 0s
+        } else {
+            hasher.update(data[*parent].as_ref());
+        }
+    }
+
+    let hash = hasher.finalize();
+    Ok(bytes_into_fr_repr_safe(hash.as_ref()).into())
 }
 
 /// Creates the encoding key.

